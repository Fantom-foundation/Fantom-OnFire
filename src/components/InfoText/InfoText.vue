<template>
    <div class="infotext">
        <p>
            Fantom on Fire monitors the total FTM burnt. <span class="number">70%</span> of the fees paid to the network
            are allocated to validators; <span class="number">15%</span> supports the <strong>GasM</strong> program,
            <span class="number">10%</span> bolsters the <strong>Ecosystem Vault</strong>, and the remaining
            <span class="number">5%</span> is permanently burnt.
        </p>
    </div>
</template>

<style lang="scss">
.infotext {
    padding: var(--f-spacer-5) var(--f-spacer-4);
}
</style>
<script setup></script>
