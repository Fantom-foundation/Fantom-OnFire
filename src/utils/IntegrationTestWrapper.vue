<script setup>
import { setupApolloProviders } from '@/plugins/apollo/apollo-providers.js';

setupApolloProviders();

defineProps(['component']);
</script>

<template>
    <component :is="component" />
</template>
