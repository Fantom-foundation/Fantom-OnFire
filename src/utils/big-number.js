import { BigNumber } from 'bignumber.js';

export const BigNumber0 = BigNumber.clone({
    DECIMAL_PLACES: 0,
    ROUNDING_MODE: BigNumber.ROUND_DOWN,
});

/**
 * @param {string|number|BigNumber} _value
 * @return {BigNumber}
 */
export function toBigNumber(_value) {
    return new BigNumber0(_value);
}

/**
 * Shift decimal point
 *
 * @param {string|number|BigNumber} _value
 * @param {number} [_dp]
 * @return {BigNumber}
 */
export function bShiftDP(_value, _dp = 18) {
    return toBigNumber(_value).shiftedBy(_dp);
}

/**
 * @param {string|number|BigNumber} value
 * @param {number} [decimals]
 * @return {BigNumber}
 */
export function bFromTokenValue(value, decimals = 18) {
    return bShiftDP(value, -decimals);
}

/**
 * @param {string|number|BigNumber} value
 * @param {number} [decimals]
 * @return {BigNumber}
 */
export function bToTokenValue(value, decimals = 18) {
    return bShiftDP(value, decimals);
}

/**
 * From Wei to Ether
 *
 * @param {string|number|BigNumber} value
 * @return {BigNumber}
 */
export function bFromWei(value) {
    return bFromTokenValue(value);
}

/**
 * From Ether to Wei
 *
 * @param {string|number|BigNumber} value
 * @return {BigNumber}
 */
export function bToWei(value) {
    return bToTokenValue(value);
}

/**
 * @param {string|number|BigNumber} _value
 * @return {string}
 */
export function toHex(_value) {
    // return web3utils.toHex(_value);
    return `0x${toBigNumber(_value).toString(16)}`;
}

/**
 * @param {string|number|BigNumber} _value
 * @return {number}
 */
export function toInt(_value) {
    return parseInt(toBigNumber(_value).toString(10));
}
