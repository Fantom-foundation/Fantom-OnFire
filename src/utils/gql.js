import { useMutation, useQuery, useResult } from '@vue/apollo-composable';

export function gqlQuery({
    query,
    variables = null,
    defaultData = null,
    pickFn = null,
    fetchPolicy = 'network-only',
    clientId = 'default',
    options = {},
}) {
    const { result, loading, error, refetch, fetchMore, onResult, onError } = useQuery(query, variables || null, {
        fetchPolicy,
        clientId,
        ...options,
    });

    const data = useResult(result, defaultData, pickFn);

    return {
        data,
        result,
        loading,
        error,
        refetch,
        fetchMore,
        onResult,
        onError,
    };
}

export function gqlMutaion({
    mutation,
    variables = null,
    fetchPolicy = 'network-only',
    clientId = 'default',
    options = {},
}) {
    const { mutate, loading, error, called, onDone, onError } = useMutation(mutation, {
        fetchPolicy,
        clientId,
        variables,
        ...options,
    });

    return {
        mutate,
        loading,
        error,
        called,
        onDone,
        onError,
    };
}
