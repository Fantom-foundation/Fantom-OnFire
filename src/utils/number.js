/**
 * Returns random positive integer between 0 and n - 1.
 *
 * @param {number} n
 * @param {boolean} randInt Return integer
 * @return {number}
 */
export function randNumber(n = 1, randInt = false) {
    // eslint-disable-next-line no-restricted-globals
    if (isNaN(parseInt(n, 10)) || n < 0) {
        throw new Error('Not a positive integer');
    }

    const value = Math.random() * n;

    if (randInt) {
        return Math.floor(value);
    }

    return value;
}

/**
 * @param {number} a
 * @param {number} b
 * @param {boolean} randInt Return integer
 * @returns {number}
 */
export function randNumberRange(a = 0, b = 1, randInt = false) {
    return a + randNumber(b - a, randInt);
}

export function randInt(n = 1) {
    return randNumber(n, true);
}

/**
 * @param {number} a
 * @param {number} b
 * @returns {number}
 */
export function randIntRange(a = 0, b = 1) {
    return randNumberRange(a, b, true);
}
