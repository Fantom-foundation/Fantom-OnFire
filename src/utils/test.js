import { beforeAll, afterAll, afterEach } from 'vitest';
import { setupNodeMockServer } from '@/mocks/node-mock-server.js';

/**
 * Get element by test id
 *
 * @param {string} id
 * @param {VueWrapper} wrapper
 * @return {*}
 */
export function getElemByTestId(id, wrapper) {
    return wrapper.find(`[data-testid="${id}"]`);
}

/**
 * Get elements by test code
 *
 * @param {string} code
 * @param {VueWrapper} wrapper
 * @return {*}
 */
export function getElemsByTestCode(code, wrapper) {
    return wrapper.findAll(`[data-testcode="${code}"]`);
}

/**
 * @param {VueWrapper} foptiongroup
 * @param {number} optionIndex
 * @return {Promise<*|undefined>}
 */
export async function checkFOptionGroupOption(foptiongroup, optionIndex = 0) {
    const inputs = foptiongroup.findAll('input');

    return inputs && inputs.length > 0 ? foptiongroup.findAll('input')[optionIndex].trigger('change') : undefined;
}

/*
export function withSetup({ composable = null, setupApollo = false } = {}) {
    let composableResult = {};

    const app = createApp({
        setup() {
            if (typeof composable === 'function') {
                composableResult = composable();

                if (setupApollo) {
                    setupApolloProviders();
                }
            }
            // suppress missing template warning
            return () => {};
        },
    });

    app.mount(document.createElement('div'));

    return { composableResult, app };
}
*/

export function getIntegrationTestWrapperMountOptions(component, options = {}) {
    const componentName = Object.keys(component)[0];

    return {
        global: { components: component },
        ...options,
        props: {
            component: componentName,
            ...(options.props || {}),
        },
    };
}

export function setupTestServer(handlers) {
    const server = setupNodeMockServer(handlers);

    function waitForRequest({ operationName = '', method = 'post', wait = 20 }) {
        return new Promise((resolve, reject) => {
            const serverEvents = server.events;
            let requestId = '';
            let response = null;

            function onRequestStart(req) {
                if (req.method.toLowerCase() === method.toLowerCase() && req.body.operationName === operationName) {
                    requestId = req.id;
                }
            }

            function onRequestMatch(req) {
                if (req.id === requestId) {
                    removeListeners();
                    setTimeout(() => {
                        resolve({ request: req, response });
                    }, wait);
                }
            }

            function onResponseMocked(res, reqId) {
                if (requestId === reqId) {
                    response = res;
                }
            }

            function onRequestUnhandled(req) {
                if (req.id === requestId) {
                    removeListeners();
                    reject(new Error(`The ${req.method} ${req.url.href} request was unhandled.`));
                }
            }

            function removeListeners() {
                serverEvents.removeListener('request:start', onRequestStart);
                serverEvents.removeListener('request:match', onRequestMatch);
                serverEvents.removeListener('request:mocked', onResponseMocked);
                serverEvents.removeListener('request:unhandled', onRequestUnhandled);
            }

            serverEvents.on('request:start', onRequestStart);
            serverEvents.on('request:match', onRequestMatch);
            serverEvents.on('response:mocked', onResponseMocked);
            serverEvents.on('request:unhandled', onRequestUnhandled);
        });
    }

    beforeAll(() => {
        server.listen({ onUnhandledRequest: 'error' });
    });
    afterAll(() => server.close());
    afterEach(() => server.resetHandlers());

    return {
        server,
        waitForRequest,
    };
}
