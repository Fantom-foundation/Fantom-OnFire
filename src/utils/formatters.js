import { bFromTokenValue } from '@/utils/big-number.js';

export const localeOptions = {
    currLocale: 'en-US',
    currency: 'USD',
    fractionDigits: 0,
    tokenPrice: 0,
};

/**
 * @param {number} _number
 * @param {number} [_fractionDigits]
 * @param {string} [_currency]
 * @param {boolean} [_variableFDigits]
 * @return {*}
 */
export function formatNumberByLocale(
    _number,
    _fractionDigits = localeOptions.fractionDigits,
    _currency,
    _variableFDigits
) {
    let options = {
        minimumFractionDigits: _variableFDigits ? 0 : _fractionDigits,
        maximumFractionDigits: _fractionDigits,
    };

    if (_currency) {
        options.style = 'currency';
        options.currency = _currency;
    }

    return new Intl.NumberFormat(localeOptions.currLocale, options).format(_number);
}

export function formatTokenValue(value, tokenPriceDecimals = 18, _fractionDigits, _currency, _variableFDigits = false) {
    return formatNumberByLocale(
        bFromTokenValue(value, tokenPriceDecimals).toNumber(),
        _fractionDigits,
        _currency ? localeOptions.currency : '',
        _variableFDigits
    );
}
