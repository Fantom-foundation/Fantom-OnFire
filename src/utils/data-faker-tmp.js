import { randIntRange, randNumberRange } from '@/utils/number.js';
import { bToTokenValue, toHex } from '@/utils/big-number.js';

function getFakeNumber(definition, def) {
    let value = def.__prv__.value;
    let step = def.__prv__.step;

    if (Array.isArray(step)) {
        step = randIntRange(step[0], step[1]);
    }

    if (definition.between) {
        value = randNumberRange(definition.between[0], definition.between[1], definition.type === 'int');
    } else if (definition.decrease) {
        value -= step;
        def.__prv__.value = value;
    } else {
        value += step;
        def.__prv__.value = value;
    }

    if (definition.decimals) {
        value = bToTokenValue(value, definition.decimals);
    }

    if (definition.toHex) {
        value = toHex(value);
    }

    return value;
}

export function getFakeData(definition = {}, len = 5) {
    const itemsDef = definition;
    const keys = Object.keys(itemsDef);
    const defs = [];
    const data = [];
    let item = {};

    keys.forEach((key) => {
        const item = itemsDef[key];
        const def = {
            name: key,
            value: null,
            __prv__: {},
        };

        if (typeof item === 'object') {
            def.__prv__.value = 'initialValue' in item ? item.initialValue : undefined;
            def.__prv__.step = item.step || 1;

            if (item.type === 'timestamp') {
                item.type = 'int';
                if (item.initialValue) {
                    def.__prv__.value = new Date(item.initialValue).valueOf() / 1000;
                }
            }

            if (item.type === 'int' || item.type === 'float') {
                if (def.__prv__.value === undefined) {
                    def.__prv__.value = 0;
                }

                def.value = getFakeNumber;
            }
        } else {
            def.value = item;
        }

        defs.push(def);
    });

    for (let i = 0; i < len; i++) {
        item = {};

        defs.forEach((def) => {
            const value = def.value;

            item[def.name] = typeof value === 'function' ? value(itemsDef[def.name], def, i) : value;
        });

        data.push(item);
    }

    return data;
}
