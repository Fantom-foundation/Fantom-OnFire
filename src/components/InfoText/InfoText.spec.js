import { mount } from '@vue/test-utils';
import { describe, it, afterEach, expect } from 'vitest';
import InfoText from '@/components/InfoText/InfoText.vue';
import { destroyWrapper } from 'fantom-vue3-components/src/test/utils.js';

function createWrapper() {
    return mount(InfoText);
}

describe('InfoText', () => {
    let wrapper = null;

    afterEach(() => {
        destroyWrapper(wrapper);
    });

    it('should match snapshot', () => {
        wrapper = createWrapper();

        expect(wrapper.element).toMatchSnapshot();
    });
});
