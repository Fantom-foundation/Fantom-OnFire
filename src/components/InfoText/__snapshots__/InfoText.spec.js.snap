// Vitest Snapshot v1

exports[`InfoText > should match snapshot 1`] = `
<div
  class="infotext"
>
  <p>
     Fantom on Fire monitors the total amount of FTM burnt, 
    <span
      class="number"
    >
      70%
    </span>
     of the fees paid to the Opera network go to validators, the remaining 
    <span
      class="number"
    >
      30%
    </span>
     are burnt forever. Fantom has a maximum supply of 
    <span
      class="number"
    >
      3.175 Billion
    </span>
    , for which the burn is slowly reducing. 
  </p>
</div>
`;
